src/crypto/test/cose.cpp

```diff
@@ -0,0 +1,219 @@
+        break;
+      case PayloadType::NestedCBOR:
+      {
+        payload.resize(1024);
+        QCBOREncodeContext ctx;
+        QCBOREncode_Init(&ctx, {payload.data(), payload.size()});
+        QCBOREncode_OpenArray(&ctx);
+        QCBOREncode_AddInt64(&ctx, 1);
+        QCBOREncode_OpenArray(&ctx);
+        QCBOREncode_AddInt64(&ctx, 2);
+        QCBOREncode_AddInt64(&ctx, 3);
+        QCBOREncode_CloseArray(&ctx);
+        QCBOREncode_CloseArray(&ctx);
+        UsefulBufC result;
+        QCBOREncode_Finish(&ctx, &result);
+        payload.resize(result.len);
+        payload.shrink_to_fit();
+      }
+      break;
+    }
+  }
+
+  std::vector<uint8_t> make_cose_sign1()
+  {
+    const auto pheaders = {
+      ccf::crypto::cose_params_int_bytes(300, value),
+      ccf::crypto::cose_params_int_int(301, 34)};
+
+    return ccf::crypto::cose_sign1(kp, pheaders, payload, false);
+  };
+
+  void verify(const std::vector<uint8_t>& cose_sign1)
+  {
+    auto verifier =
+      ccf::crypto::make_cose_verifier_from_key(kp.public_key_pem());
+    if (detached_payload)
+    {
+      verifier->verify_detached(cose_sign1, payload);
+    }
+    else
+    {
+      std::span<uint8_t> payload_;
+      REQUIRE(verifier->verify(cose_sign1, payload_));
+      std::vector<uint8_t> payload_copy(payload_.begin(), payload_.end());
+      REQUIRE(payload == payload_copy);
+    }
+  };
+};
+
+TEST_CASE("Verification and payload invariant")
+{
+  for (auto type :
+       {PayloadType::Detached, PayloadType::Flat, PayloadType::NestedCBOR})
+  {
+    Signer signer(type);
+    auto csp = signer.make_cose_sign1();
+    signer.verify(csp);
+
+    for (const auto& key : keys)
+    {
+      for (const auto& position : positions)
+      {
+        auto csp_set =
+          ccf::cose::edit::set_unprotected_header(csp, key, position, value);
+
+        signer.verify(csp_set);
+      }
+    }
+  }
+}
+
+TEST_CASE("Idempotence")
+{
+  for (auto type :
+       {PayloadType::Detached, PayloadType::Flat, PayloadType::NestedCBOR})
+  {
+    Signer signer(type);
+    auto csp = signer.make_cose_sign1();
+
+    for (const auto& key : keys)
+    {
+      for (const auto& position : positions)
+      {
+        auto csp_set_once =
+          ccf::cose::edit::set_unprotected_header(csp, key, position, value);
+
+        auto csp_set_twice = ccf::cose::edit::set_unprotected_header(
+          csp_set_once, key, position, value);
+        REQUIRE(csp_set_once == csp_set_twice);
+      }
+    }
+  }
+}
+
+TEST_CASE("Check unprotected header")
+{
+  for (auto type :
+       {PayloadType::Detached, PayloadType::Flat, PayloadType::NestedCBOR})
+  {
+    Signer signer(type);
+    auto csp = signer.make_cose_sign1();
+
+    for (const auto& key : keys)
+    {
+      for (const auto& position : positions)
+      {
+        auto csp_set =
+          ccf::cose::edit::set_unprotected_header(csp, key, position, value);
+
+        std::vector<uint8_t> ref(1024);
```

Comment:
I understand these are tests, but the loop body is probably worth splitting to a couple of functions

Reply:
If they were used somewhere else I would agree, but here...